modules/geoedgeRtdProvider.js

```diff
@@ -0,0 +1,213 @@
+ * @module modules/geoedgeProvider
+ * @requires module:modules/realTimeData
+ */
+
+/**
+ * @typedef {Object} ModuleParams
+ * @property {string} key
+ * @property {?Object} bidders
+ * @property {?boolean} wap
+ * @property {?string} keyName
+ */
+
+import { submodule } from '../src/hook.js';
+import { ajax } from '../src/ajax.js';
+import { generateUUID, insertElement, isEmpty, logError } from '../src/utils.js';
+
+/** @type {string} */
+const SUBMODULE_NAME = 'geoedge';
+/** @type {string} */
+export const WRAPPER_URL = '//wrappers.geoedge.be/wrapper.html';
+/** @type {string} */
+/* eslint-disable no-template-curly-in-string */
+export const HTML_PLACEHOLDER = '${creative}';
+/** @type {string} */
+const PV_ID = generateUUID();
+/** @type {string} */
+const HOST_NAME = '//rumcdn.geoedge.be';
+/** @type {string} */
+const FILE_NAME = 'grumi.js';
+/** @type {function} */
+export let getClientUrl = (key) => `${HOST_NAME}/${key}/${FILE_NAME}`;
+/** @type {string} */
+export let wrapper
+/** @type {boolean} */;
+let wrapperReady;
+/** @type {boolean} */;
+let preloaded;
+
+/**
+ * fetches the creative wrapper
+ * @param {function} sucess - success callback
+ */
+export function fetchWrapper(success) {
+  if (wrapperReady) {
+    return success(wrapper);
+  }
+  ajax(WRAPPER_URL, success);
+}
+
+/**
+ * sets the wrapper and calls preload client
+ * @param {string} responseText
+ */
+export function setWrapper(responseText) {
+  wrapperReady = true;
+  wrapper = responseText;
+}
+
+/**
+ * preloads the client
+  * @param {string} key
+ */
+export function preloadClient(key) {
+  let link = document.createElement('link');
+  link.rel = 'preload';
+  link.as = 'script';
+  link.href = getClientUrl(key);
+  link.onload = () => { preloaded = true };
+  insertElement(link);
+}
+
+/**
+ * creates identity function for string replace without special replacement patterns
+ * @param {string} str
+ * @return {function}
+ */
+function replacer(str) {
+  return function () {
+    return str;
+  }
+}
+
+export function wrapHtml(wrapper, html) {
+  return wrapper.replace(HTML_PLACEHOLDER, replacer(html));
+}
+
+/**
+ * generate macros dictionary from bid response
+ * @param {Object} bid
+ * @param {string} key
+ * @return {Object}
+ */
+function getMacros(bid, key) {
+  return {
+    '${key}': key,
+    '%%ADUNIT%%': bid.adUnitCode,
+    '%%WIDTH%%': bid.width,
+    '%%HEIGHT%%': bid.height,
+    '%%PATTERN:hb_adid%%': bid.adId,
+    '%%PATTERN:hb_bidder%%': bid.bidderCode,
+    '%_isHb!': true,
+    '%_hbcid!': bid.creativeId || '',
+    '%%PATTERN:hb_pb%%': bid.pbHg,
+    '%%SITE%%': location.hostname,
+    '%_pimp%': PV_ID
+  };
+}
+
+/**
+ * replace macro placeholders in a string with values from a dictionary
+ * @param {string} wrapper
+ * @param {Object} macros
+ * @return {string}
+ */
+function replaceMacros(wrapper, macros) {
+  let key;
+  for (key in macros) {
+    wrapper = wrapper.replace(key, replacer(macros[key]));
+  }
+  return wrapper;
+}
```

Comment:
I see that you're looping for every key in the map. Would it be possible to use RegEx and loop only once? Maybe something like the code below would work?

```(javascript)
function replaceAll(str,mapObj){
    var re = new RegExp(Object.keys(mapObj).join("|"),"gi");

    return str.replace(re, function(matched){
        return mapObj[matched.toLowerCase()];
    });
}
```

Reply:
It is possible, we only needed to add escape character in the beginning for `${key}' because of the opening `$`.